pallets/gas/src/property_tests/strategies.rs

```diff
@@ -0,0 +1,84 @@
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use proptest::prelude::*;
+
+const MAX_ACTIONS: usize = 1000;
+
+/// Enum representing gas tree functions that mutate the state of the existing gas tree.
+///
+/// Instead of `H256` values, `usize` values are used to represent ids of nodes in the tree.
+/// These ids are like **handles** pointing to some existing id in the tree. It's used the next way:
+/// ```no_run
+/// let node_ids: Vec<H256> = Vec::new();
+///
+/// // ...
+/// let action = GasTreeAction::Split(12312312);
+///
+/// // ...
+///
+/// if let GasTreeAction::Split(parent_idx) = action {
+///     // To understand how `ring_get` works see `RingGet` traits implementation.
+///     let parent_id = node_ids.ring_get(parent_idx).unwrap();
+///     Gas::split(parent_id, H256::random())
+/// }
+/// ```
+///
+#[derive(Debug, Clone, Copy)]
+pub(super) enum GasTreeAction {
+    /// Performs split of the node under the bound `usize` index.
+    Split(usize),
+    /// Performs split of the node under the bound `usize` index with `u64` amount.
+    SplitWithValue(usize, u64),
+    /// Spends `u64` amount of value from the node under the bound `usize` index.
+    Spend(usize, u64),
+    /// Consumes the node under the bound `usize` index.
+    Consume(usize),
+}
+
+/// Returns random vector of `GasTreeAction`s with a tree's root max balance.
+///
+/// Execution of the random set of `GasTreeAction`s results in a unique gas tree in the
+/// storage, which is needed to perform property tests. Max balance sets upper boundary on the
+/// amount by which node's value can be decreased (in split_with_value and spend procedures).
+/// Also max balance defines root's balance.  
+pub(super) fn gas_tree_props_test_strategy() -> impl Strategy<Value = (u64, Vec<GasTreeAction>)> {
+    any::<u64>()
+        .prop_flat_map(|max_balance| (Just(max_balance), gas_tree_action_strategy(max_balance)))
+}
+
+/// Generates random vector of `GasTreeAction`s that defines how gas tree is created.
+pub(super) fn gas_tree_action_strategy(
+    max_balance: u64,
+) -> impl Strategy<Value = Vec<GasTreeAction>> {
+    let action_random_variant = prop_oneof![
+        (any::<usize>(), 0..max_balance).prop_flat_map(|(id, amount)| {
+            prop_oneof![
+                Just(GasTreeAction::SplitWithValue(id, amount)),
+                Just(GasTreeAction::Spend(id, amount))
+            ]
+        }),
+        any::<usize>().prop_flat_map(|id| {
+            prop_oneof![
+                Just(GasTreeAction::Consume(id)),
+                Just(GasTreeAction::Split(id))
+            ]
+        }),
```

Comment:
Isn't this randomiser generate them chaotically, skipping checks that we cannot consume twice, for example?

Reply:
All in all it creates a random set of `GasTreeActions`, where ids of `GasTreeActions::Consume` variant can point to the same id, i.e. performing consume twice check.